pkg/controller/prpqr_reconciler/prpqr_reconciler.go

```diff
@@ -0,0 +1,193 @@
+		if !ok {
+			logrus.WithField("type", fmt.Sprintf("%T", o)).Error("Got object that was not a PullRequestPayloadQualificationRun")
+			return nil
+		}
+
+		return []reconcile.Request{
+			{NamespacedName: types.NamespacedName{Namespace: prpqr.Namespace, Name: prpqr.Name}},
+		}
+	})
+}
+
+type reconciler struct {
+	logger *logrus.Entry
+	client ctrlruntimeclient.Client
+}
+
+func (r *reconciler) Reconcile(ctx context.Context, req reconcile.Request) (reconcile.Result, error) {
+	logger := r.logger.WithField("request", req.String())
+	err := r.reconcile(ctx, req, logger)
+	if err != nil {
+		logger.WithError(err).Error("Reconciliation failed")
+	} else {
+		logger.Info("Finished reconciliation")
+	}
+	return reconcile.Result{}, controllerutil.SwallowIfTerminal(err)
+}
+
+func (r *reconciler) reconcile(ctx context.Context, req reconcile.Request, logger *logrus.Entry) error {
+	logger = logger.WithField("namespace", req.Namespace).WithField("prpqr_name", req.Name)
+	logger.Info("Starting reconciliation")
+
+	prpqr := &v1.PullRequestPayloadQualificationRun{}
+	if err := r.client.Get(ctx, ctrlruntimeclient.ObjectKey{Namespace: req.Namespace, Name: req.Name}, prpqr); err != nil {
+		return fmt.Errorf("failed to get the PullRequestPayloadQualificationRun: %s in namespace %s: %w", req.Name, req.Namespace, err)
+	}
+
+	prowjobs := generateProwjobs(prpqr.Spec.PullRequest.Org, prpqr.Spec.PullRequest.Repo, prpqr.Spec.PullRequest.BaseRef, req.Name, req.Namespace)
+	for _, pj := range prowjobs {
+		logger = logger.WithFields(logrus.Fields{"name": pj.Name, "namespace": req.Namespace})
+
+		err := r.client.Get(ctx, ctrlruntimeclient.ObjectKey{Namespace: pj.Namespace, Name: pj.Name}, &prowv1.ProwJob{})
+		if err != nil && !kerrors.IsNotFound(err) {
+			return fmt.Errorf("failed to get the PullRequestPayloadQualificationRun: %s in namespace %s: %w", req.Name, req.Namespace, err)
+		}
+
+		if !kerrors.IsNotFound(err) {
+			continue
+		}
+
+		logger.Info("Creating prowjob...")
+		if err := r.client.Create(ctx, &pj); err != nil {
+			return fmt.Errorf("failed to create prowjob: %w", err)
+		}
+
+		// There is some delay until it gets back to our cache, so block until we can retrieve
+		// it successfully.
+		key := ctrlruntimeclient.ObjectKey{Namespace: pj.Namespace, Name: pj.Name}
+		if err := wait.Poll(100*time.Millisecond, 5*time.Second, func() (bool, error) {
+			if err := r.client.Get(ctx, key, &prowv1.ProwJob{}); err != nil {
+				if kerrors.IsNotFound(err) {
+					return false, nil
+				}
+				return false, fmt.Errorf("getting prowJob failed: %w", err)
+			}
+			return true, nil
+		}); err != nil {
+			return fmt.Errorf("failed to wait for created ProwJob to appear in cache: %w", err)
+		}
+
+		prpqr.Status.Jobs = append(prpqr.Status.Jobs, v1.PullRequestPayloadJobStatus{
+			ReleaseJobName: pj.Spec.Job,
+			ProwJob:        pj.Name,
+			Status:         pj.Status,
+		})
+
+		logger.Info("Updating PullRequestPayloadQualificationRun...")
```

Comment:
Could we do this via the another controller? (if the another one watch the create events of prowjob)

Reply:
The other controller watches only the updates. I could have made it watch the create events as well, and update the `prpqr` but this way we lock the behavior of each controller.

Reply:
Controller1 watches PullRequestPayloadQualificationRun and creates ProwJobs if needed.
Controller2 watches ProwJobs and updates PullRequestPayloadQualificationRun's status accordingly.

How they get locked? I do not see the infinite loop either.

Reply:
I mean that each controller's behavior is related to each action. The first controller is responsible to watch only `create` events and creates the corresponding pj and updates the prpqr status. The second controller is watching only `update` events for prowjobs that hold the label, and then updates the prpqr job statuses.